editoast/src/views/pathfinding/catenaries.rs

```diff
@@ -0,0 +1,473 @@
+use crate::error::Result;
+use crate::infra::Infra;
+use crate::infra_cache::InfraCache;
+use crate::models::{pathfinding::Pathfinding, Retrieve};
+use crate::schema::{Direction, ObjectType};
+use crate::views::pathfinding::PathfindingError;
+use crate::DbPool;
+use actix_web::{
+    dev::HttpServiceFactory,
+    get, services,
+    web::{block, Data, Json, Path},
+};
+use chashmap::CHashMap;
+use rangemap::RangeMap;
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::ops::Range;
+
+pub fn routes() -> impl HttpServiceFactory {
+    services![catenaries_on_path,]
+}
+
+/// A struct used for the result of the catenaries_on_path endpoint
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+struct CatenaryRange {
+    begin: f64,
+    end: f64,
+    mode: String,
+}
+
+impl CatenaryRange {
+    fn list_from_range_map(range_map: &RangeMap<Float, String>) -> Vec<CatenaryRange> {
+        range_map
+            .iter()
+            .map(|(range, mode)| CatenaryRange {
+                begin: range.start.0,
+                end: range.end.0,
+                mode: mode.to_string(),
+            })
+            .collect()
+    }
+}
+
+/// A struct to make f64 Ord, to use in RangeMap
+#[derive(Debug, PartialEq, Copy, Clone, Serialize)]
+struct Float(f64);
+
+impl Float {
+    fn new(f: f64) -> Self {
+        assert!(f.is_finite());
+        Self(f)
+    }
+}
+
+impl From<f64> for Float {
+    fn from(f: f64) -> Self {
+        Self::new(f)
+    }
+}
+
+impl Ord for Float {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0.partial_cmp(&other.0).unwrap()
+    }
+}
+
+impl PartialOrd for Float {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+impl Eq for Float {}
+
+/// Build a rangemap for each track section, giving the voltage for each range
+fn map_catenary_modes(
+    infra_cache: &InfraCache,
+    track_ids: HashSet<String>,
+) -> HashMap<String, RangeMap<Float, String>> {
+    let unique_catenary_ids = track_ids
+        .iter()
+        .flat_map(|track_id| {
+            infra_cache
+                .get_track_refs_type(track_id, ObjectType::Catenary)
+                .into_iter()
+        })
+        .map(|catenary_ref| &catenary_ref.obj_id)
+        .collect::<HashSet<_>>();
```

Comment:
In my experience the fastest way to dedup a collection in rust is not using an hashset.
You collect into a `Vec`. Then `Vec`, then sort it and call the [dedup](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.dedup) function.

Here is an example:

https://github.com/flomonster/kdtree-ray/blob/afdf6562b5a72823eacbec4dd415fb3da698bed7/src/kdtree.rs#L58-L87

Notes:
- Doing this might be overkill here (You can ignore this comment if the endpoint response takes less than a few milliseconds) 
- Otherwise, it can be interesting to benchmark this.

Reply:
With the --release flag, the endpoint takes between 15 and 20ms. I tried your suggested change and it did not affect the computation time